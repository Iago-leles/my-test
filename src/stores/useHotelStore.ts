import { defineStore } from 'pinia'
import { ref, computed } from 'vue'

import type { Hotel } from '@/types'

import hotelData from '@/db/hotel.json'

import { HotelOrder } from '@/enums/hotels'

export const useHotelStore = defineStore('hotel', () => {
  const hotels = ref<Hotel[]>(hotelData.flatMap((data) => data.hotels) as Hotel[])
  const selectedHotel = ref<Hotel | null>(null)
  const isLoading = ref(false)
  const currentPage = ref(1)
  const itemsPerPage = 10
  const selectedPlace = ref<{ value: string; label: string } | null>(null)
  const orderBy = ref(HotelOrder.RECOMMENDED)
  const orderOptions = [HotelOrder.RECOMMENDED, HotelOrder.BEST_RATED]
  const filterName = ref('')

  const allHotels = computed(() => {
    return hotelData
  })

  function setFilterName(value: string) {
    filterName.value = value
  }

  const filteredHotels = computed(() => {
    let result = hotelData.flatMap((data) => data.hotels || []) as Hotel[]

    if (selectedPlace.value?.value) {
      result = allHotels.value
        .filter((places) => places.placeId === Number(selectedPlace.value?.value))
        .flatMap((places) => places.hotels) as Hotel[]
    }

    if (filterName.value.trim()) {
      const searchTerm = filterName.value.trim().toLowerCase()
      result = result.filter((hotel) => hotel.name.toLowerCase().includes(searchTerm))
    }

    const sortedOptions: Record<HotelOrder, Hotel[]> = {
      [HotelOrder.RECOMMENDED]: [...result].sort((a, b) => Number(a.price) - Number(b.price)),
      [HotelOrder.BEST_RATED]: [...result].sort((a, b) => Number(b.stars) - Number(a.stars)),
    }

    return sortedOptions[orderBy.value]
  })

  const displayedHotels = computed(() => {
    const endIndex = currentPage.value * itemsPerPage
    return filteredHotels.value.slice(0, endIndex)
  })

  function fetchNextPage(index: number, done: () => void) {
    setTimeout(() => {
      if (displayedHotels.value.length < filteredHotels.value.length) {
        currentPage.value++
        done()

        return
      }

      done()
    }, 500)
  }

  function setSelectedPlace(place: { value: string; label: string } | null) {
    selectedPlace.value = place || null
  }

  function setSelectedHotel(hotel: Hotel | null) {
    selectedHotel.value = hotel || null
  }

  return {
    hotels,
    isLoading,
    filterName,
    setFilterName,
    filteredHotels,
    displayedHotels,
    selectedPlace,
    orderOptions,
    orderBy,
    selectedHotel,
    fetchNextPage,
    setSelectedPlace,
    setSelectedHotel,
  }
})
