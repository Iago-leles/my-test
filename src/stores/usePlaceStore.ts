import { defineStore } from 'pinia'
import { ref, computed } from 'vue'

import type { Place } from 'src/types'

import placesData from 'src/db/place.json'

export const usePlaceStore = defineStore('place', () => {
  const places = ref<Place[]>(placesData)
  const selectedPlace = ref<{ label: string; value: string } | null>(null)
  const hasSearched = ref(false)

  const formattedPlaces = computed(() => {
    return places.value.map((place) => ({
      label: `${place.name}, ${place.state.name}`,
      value: String(place.placeId),
    }))
  })

  const selectedPlaceData = computed(() => {
    return formattedPlaces.value.find((place) => place.value === selectedPlace.value?.value)
  })

  function setSelectedPlace(place: { label: string; value: string } | null) {
    selectedPlace.value = place
  }

  function setHasSearched(value: boolean) {
    hasSearched.value = value
  }

  function resetSearch() {
    hasSearched.value = false
  }

  return {
    places,
    selectedPlace,
    hasSearched,
    formattedPlaces,
    selectedPlaceData,
    setSelectedPlace,
    setHasSearched,
    resetSearch,
  }
})
