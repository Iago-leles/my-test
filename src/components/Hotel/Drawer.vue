<template>
  <q-drawer
    overlay
    v-model="isHotelSelected"
    side="right"
    bordered
    class="hotel-drawer"
    :width="$q.screen.lt.md ? 400 : 500"
    @hide="onHide"
  >
    <q-scroll-area class="fit" :style="$q.screen.lt.md ? 'width: 88vw !important' : ''">
      <div v-if="$q.screen.lt.md" class="row justify-end" @click="onHide">
        <q-icon name="close" size="24px" class="q-mt-none text-grey-8" />
      </div>

      <span class="row items-center q-mt-none">
        <div class="text-grey-6 q-mr-xs text-caption">
          {{ stars.toFixed(1) }}
        </div>

        <q-rating v-model="stars" size="1.2em" color="amber" readonly />
      </span>

      <div>
        <p class="text-h5 text-grey-8 q-mb-none">{{ hotel?.name }}</p>

        <p class="text-body2 text-grey-8">{{ hotel?.address.fullAddress }}</p>
      </div>

      <app-carousel :images="hotel?.images || []" :autoplay="true" show-counter full-rounded />

      <div class="q-mt-md">
        <div v-if="shouldRenderAmenities">
          <p class="text-h6 q-mb-xs text-grey-8 text-weight-light">Facilidades do hotel</p>

          <q-separator size="2px" />

          <div class="row q-mt-md">
            <div
              v-if="shouldRenderAmenities"
              class="row q-mt-none"
              style="display: flex; justify-content: space-between; width: 100%"
            >
              <div
                v-for="amenity in displayedAmenities"
                :key="amenity?.label || ''"
                class="amenity-item q-pa-xs text-grey-8"
              >
                <q-icon :name="amenity?.icon" size="16px" class="q-mr-sm" />

                <span class="text-subtitle2 text-weight-light">{{ amenity?.label }}</span>
              </div>
            </div>
          </div>
        </div>

        <div class="row justify-center q-my-xs">
          <q-btn
            v-if="hasMoreAmenities"
            color="blue"
            outline
            rounded
            class="q-mb-md"
            @click="showAllAmenities = !showAllAmenities"
          >
            {{ showAllAmenities ? 'Mostrar menos' : 'Mostrar todas as facilidades' }}
          </q-btn>
        </div>

        <q-separator size="2px" v-if="hotel?.description && shouldRenderAmenities" />

        <div class="q-mt-md" v-if="hotel?.description">
          <p class="text-h6 q-mb-md text-weight-light">Conhe√ßa um pouco mais</p>

          <p class="text-body2 text-grey-8" v-html="hotel?.description" />
        </div>
      </div>
    </q-scroll-area>
  </q-drawer>
</template>

<script setup lang="ts">
import { computed, ref } from 'vue'

import type { Hotel } from '@/types'

import { useHotelStore } from '@/stores/useHotelStore'

import AppCarousel from '@/components/ui/AppCarousel.vue'

import { amenitiesIcons } from '@/static/hotels'

import type { HotelAmenities } from '@/enums/hotels'

const props = defineProps<{
  hotel?: Hotel | null
}>()

const hotelStore = useHotelStore()

const isHotelSelected = computed(() => hotelStore.selectedHotel !== null)

const stars = computed(() => Number(props.hotel?.stars))

const hotelAmenities = computed(() => {
  return props.hotel?.amenities
    ?.filter((amenity) => amenity.key && amenity.label)
    .map((amenity) => {
      return {
        label: amenity.label,
        icon: amenitiesIcons[amenity.key as HotelAmenities],
      }
    })
})

const shouldRenderAmenities = computed(
  () => Array.isArray(props.hotel?.amenities) && props.hotel.amenities.length > 0,
)

const showAllAmenities = ref(false)
const INITIAL_AMENITIES_COUNT = 3

const displayedAmenities = computed(() => {
  if (!hotelAmenities.value) return []
  return showAllAmenities.value
    ? hotelAmenities.value
    : hotelAmenities.value.slice(0, INITIAL_AMENITIES_COUNT)
})

const hasMoreAmenities = computed(() => {
  return hotelAmenities.value && hotelAmenities.value.length > INITIAL_AMENITIES_COUNT
})

const onHide = () => {
  hotelStore.setSelectedHotel(null)
}
</script>

<style lang="scss">
.hotel-drawer {
  padding: 16px;
  width: 500px !important;
}

.amenity-item {
  display: flex;
  align-items: center;
  min-height: 32px;
}
</style>
