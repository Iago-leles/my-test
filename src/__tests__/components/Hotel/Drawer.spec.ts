import { mount, flushPromises } from '@vue/test-utils'
import { expect, describe, it, vi } from 'vitest'
import { createTestingPinia } from '@pinia/testing'
import { useHotelStore } from '@/stores/useHotelStore'

import Drawer from '@/components/Hotel/Drawer.vue'
import type { Hotel } from 'src/types'

describe('Drawer', () => {
  it('should match snapshot', async () => {
    const hotel = {
      id: 1,
      name: 'Hotel Test',
      address: { fullAddress: '123 Test St' },
      stars: 4.5,
      price: 100,
      amenities: [],
      images: [],
      hasRefundableRoom: true,
      description: 'Descrição do hotel',
    }

    const wrapper = mount(Drawer, {
      props: { hotel: hotel as unknown as Hotel },
      global: {
        plugins: [
          createTestingPinia({
            createSpy: vi.fn,
            fakeApp: true,
          }),
        ],
      },
    })

    await flushPromises()

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should set selected hotel to null on hide', () => {
    const hotel = {
      id: 1,
      name: 'Hotel Test',
      address: { fullAddress: '123 Test St' },
      stars: 4.5,
      price: 100,
      amenities: [
        {
          key: 'WI_FI',
          label: 'Internet',
        },
        {
          key: 'POOL',
          label: 'Piscina',
        },
        {
          key: 'PARKING',
          label: 'Estacionamento',
        },
        {
          key: 'RESTAURANT',
          label: 'Restaurante',
        },
      ],
      images: [],
      hasRefundableRoom: true,
      description: 'Descrição do hotel',
    }

    const pinia = createTestingPinia({
      createSpy: vi.fn,
      stubActions: false,
    })

    const wrapper = mount(Drawer, {
      props: { hotel: hotel as unknown as Hotel },
      global: {
        plugins: [pinia],
      },
    })

    wrapper.vm.$emit('hide')

    const hotelStore = useHotelStore(pinia)
    expect(hotelStore.selectedHotel).toBeNull()
  })
})
