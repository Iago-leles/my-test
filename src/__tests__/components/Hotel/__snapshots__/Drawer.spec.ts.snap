// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Drawer > should match snapshot 1`] = `""`;
